cmd/segment-reaper/observer.go

```diff
@@ -143,11 +156,87 @@ func (obsvr *observer) processSegment(ctx context.Context, path metainfo.ScopedP
 // analyzeProject analyzes the objects in obsv.objects field for detecting bad
 // segments and writing them to objs.writer.
 func (obsvr *observer) analyzeProject(ctx context.Context) error {
-	// TODO this part will be implemented in next PR
-	// TODO(if): For what is this?
+	for bucket, objects := range obsvr.objects {
+		for path, object := range objects {
+			if object.skip {
+				continue
+			}
+
+			brokenObject := false
+			if object.hasLastSegment {
+				if object.expectedNumberOfSegments == 0 {
+					if !object.segments.IsSequence() {
+						brokenObject = true
+					}
+				} else if object.segments.Count() != int(object.expectedNumberOfSegments)-1 {
+					// expectedNumberOfSegments-1 because 'segments' doesn't contain last segment
+					brokenObject = true
+				}
+
+				if brokenObject {
+					err := obsvr.printSegment(ctx, "l", bucket, path)
+					if err != nil {
+						return err
+					}
+				}
+			} else {
+				brokenObject = true
+			}
+
+			if brokenObject {
+				for index := 0; index < maxNumOfSegments; index++ {
+					has, err := object.segments.Has(index)
+					if err != nil {
+						panic(err)
+					}
+					if has {
+						err := obsvr.printSegment(ctx, "s"+strconv.Itoa(index), bucket, path)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
 	return nil
 }
 
+func (obsvr *observer) printSegment(ctx context.Context, segmentIndex, bucket, path string) error {
+	creationDate, err := pointerCreationDate(ctx, obsvr.db, obsvr.lastProjectID, segmentIndex, bucket, path)
+	if err != nil {
+		return err
+	}
+	encodedPath := base64.StdEncoding.EncodeToString([]byte(path))
+	err = obsvr.writer.Write([]string{
+		obsvr.lastProjectID,
+		segmentIndex,
+		bucket,
+		encodedPath,
+		creationDate,
+	})
+	if err != nil {
+		return err
+	}
+	obsvr.zombieSegments++
```

Comment:
Is there a reason this increment should be done here as opposed to in the `analyzeProject` method?

Reply:
in `analyzeProject` I would need to increment in two places, here its only one place.

Reply:
fair enough